stereo_image_proc/src/stereo_image_proc/disparity_fpga.cpp

```diff
@@ -0,0 +1,399 @@
+  // int left = block_matcher_.getDisparityRange() + block_matcher_.getMinDisparity() + border - 1;
+  // int wtf;
+  /* if (block_matcher_.getMinDisparity() >= 0) {
+     wtf = border + block_matcher_.getMinDisparity();
+  } else {
+     wtf = std::max(border, -block_matcher_.getMinDisparity());
+  } */
+  // int right = disp_msg->image.width - 1 - wtf;
+  // int top = border;
+  // int bottom = disp_msg->image.height - 1 - border;
+  // disp_msg->valid_window.x_offset = left;
+  // disp_msg->valid_window.y_offset = top;
+  // disp_msg->valid_window.width = right - left;
+  // disp_msg->valid_window.height = bottom - top;
+
+  // Vitis Vision library ROS to CV image
+  cv_bridge::CvImagePtr cv_ptr_l;
+  cv_bridge::CvImagePtr cv_ptr_r;
+  cv::Mat hls_disp16;
+  static const double inv_dpp = 1.0 / NO_OF_DISPARITIES;
+
+  try {
+  cv_ptr_l = cv_bridge::toCvCopy(l_image_msg, sensor_msgs::image_encodings::MONO8);
+  cv_ptr_r = cv_bridge::toCvCopy(r_image_msg, sensor_msgs::image_encodings::MONO8);
+  } catch (cv_bridge::Exception & e) {
+	  RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
+  /*        TRACEPOINT(stereo_image_proc_disparity_cv,
+	  static_cast<const void *>(this),
+	  static_cast<const void *>(&(*l_image_msg)),
+	  static_cast<const void *>(&(*r_image_msg)));*/
+    return;
+  }
+
+  // OpenCL section WIP
+
+  cl_int err;
+  size_t image_l_in_size_bytes, image_r_in_size_bytes, image_out_size_bytes;
+
+  // Assume MONO (1 channels)
+  hls_disp16.create(cv::Size(l_info_msg->width,
+                                l_info_msg->height), CV_16UC1);
+
+  image_l_in_size_bytes = l_info_msg->height * l_info_msg->width *
+                                  1 * sizeof(unsigned char);
+
+  // Assume same size as left image for the buffer allocation
+  image_r_in_size_bytes = image_l_in_size_bytes;
+
+  image_out_size_bytes = l_info_msg->height * l_info_msg->width *
+	                          1 * sizeof(unsigned char);
+
+
+  // Allocate the buffers:
+  OCL_CHECK(err, cl::Buffer imageLToDevice(*context_, CL_MEM_READ_ONLY,
+                                          image_l_in_size_bytes, NULL, &err));
+  OCL_CHECK(err, cl::Buffer imageRToDevice(*context_, CL_MEM_READ_ONLY,
+                                          image_r_in_size_bytes, NULL, &err));
+  OCL_CHECK(err, cl::Buffer imageFromDevice(*context_, CL_MEM_WRITE_ONLY,
+                                            image_out_size_bytes, NULL, &err));
+  // Set the kernel arguments
+  OCL_CHECK(err, err = krnl_->setArg(0, imageLToDevice));
+  OCL_CHECK(err, err = krnl_->setArg(1, imageRToDevice));
+  OCL_CHECK(err, err = krnl_->setArg(3, imageFromDevice));
+  OCL_CHECK(err, err = krnl_->setArg(2, bm_state_));
+  OCL_CHECK(err, err = krnl_->setArg(4, disp_info_msg->width));
+  OCL_CHECK(err, err = krnl_->setArg(5, disp_info_msg->height));
+
+  OCL_CHECK(err, queue_->enqueueWriteBuffer(imageRToDevice, CL_TRUE, 0, image_r_in_size_bytes, cv_ptr_r->image.data));
```

Comment:
I think enqueuing left and right buffers can be done in parallel, so no need to block the first
Simply make sure to flush the queue before enqueue the stereo matching kernel task